Fixed win screen,  new game button and got my new game function to work again

prra-codes opened the pull request:
Summary: Created useEffect to fix the win screens and new game button. Adjusted my newGame function to make an axios post request, and the request on the server resets the boardArr and currentColumn variables on the server side. The useEffect that runs on the client side every 0.5 seconds will then update the client side board and currentColumns states.

Test Plan: Ran npm start and npm run dev. Things are looking good!

prra-codes committed:
dealing with I believe to be  async issues

Summary: Three main issues. 1. Win screen only shows a move after the winning move has been made (one move after a Connect 4 has been created).
2. New Game button shows after two moves have been on the board, rather than just one.
3. I'm having difficulty creating a newGame func on the server side, so that I can restart the board game again.

Test Plan: Ran npm start and npm run dev. 1. and 2. are the biggest issues, as I am uncertain of the cause.

prra-codes committed:
fixed win screen, new game button and put new game function on server

summary: created useEffect to fix win screen and new game button. Adjusted my newGame function to send variables to the server and update the variables to reset the newBoard, column and player states in the server. used res.status(200) and res.send() to update client for newGame function.

test plan: ran npm start and npm run dev. used react components in google dev tools to examine which states were changing before making further adjustments. Tested the game. seems to work properly.

jacksongabbard commented on client/src/components/Board/Board.js:
```diff
@@ -64,9 +68,30 @@ const Board = () => {
+    await axios.post("http://localhost:8000/newGame", {
+      newBoard,
+      resetCurrentColumns,
+      player,
```
Why do you need to send this info to the server? If you know you want to create a newGame, you should be able to do 100% of the setup on the server without the client passing specific game state data to the server.

prra-codes replied:
Understood.

jacksongabbard commented on server/server.js:
```diff
@@ -40,13 +35,35 @@ app.post("/setPiece", async (req, res) => {
+});
+
+app.post("/newGame", async (req, res) => {
+  const newGameData = req.body;
+
+  const resetCurrentColumns = newGameData["resetCurrentColumns"];
+
+  currentColumns = [...resetCurrentColumns];
+
+  const resetGameBoard = newGameData["newBoard"];
+  boardArr = [...resetGameBoard];
+
+  const player = newGameData["player"];
+
+  currentPlayer = player;
```
This overall change is good, except this part. You shouldn't accept the entire game state blindly from the user. Suppose they send you a game state that is an array with 1,000,000 rows and a 1,000,000 columns to see if they can crash your server?

prra-codes replied:
I'll update it. Thank you.

jacksongabbard reviewed (changes requested):
You need to change the way the new game functionality works.

Also, you need to update the title and the description of the pull request to describe what it actually is now. "dealing with I believe to be async issues" isn't a very useful title for future you to make sense of.

prra-codes committed:
Adjusted new game function.

Summary: removed resetCurrentColumns and newBoard variables from client and add them to the server, so the server resets the boardArr and currentColumn variables.

Test Plan: Ran npm start and npm run dev, seems to work fine.

jacksongabbard commented on server/server.js:
```diff
@@ -40,13 +35,36 @@ app.post("/setPiece", async (req, res) => {
+  currentColumns = [...resetCurrentColumns];
+
+  boardArr = [...newBoard];
+
```
Do you need to reset who the current player is?

prra-codes replied:
No, when a player wins, the current player is already changed due to the setPiece function on the client and server.

E.g. If it's turn red, and red makes the winning move, the cursor will turn yellow and the board game will say "RED WINS!". So when the new game button is clicked, the cursor is already yellow.

jacksongabbard reviewed (approved):
One question down below, but otherwise this looks good to me.

prra-codes commented:
Thank you!

No, when a player wins, the current player is already changed due to the
setPiece function on the client and server.

E.g. If it's turn red, and red makes the winning move, the cursor will turn
yellow and the board game will say "RED WINS!". So when the new game button
is clicked, the cursor is already yellow.

On Thu, Jul 6, 2023 at 11:40 AM Jackson Gabbard ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> One question down below, but otherwise this looks good to me.
> ------------------------------
>
> In server/server.js
> <https://github.com/prra-codes/peter-arthur-portfolio/pull/8#discussion_r1254259168>
> :
>
> > +app.post("/newGame", async (req, res) => {
> +  const newBoard = [
> +    [null, null, null, null, null, null, null],
> +    [null, null, null, null, null, null, null],
> +    [null, null, null, null, null, null, null],
> +    [null, null, null, null, null, null, null],
> +    [null, null, null, null, null, null, null],
> +    [null, null, null, null, null, null, null],
> +  ];
> +
> +  const resetCurrentColumns = [5, 5, 5, 5, 5, 5, 5];
> +
> +  currentColumns = [...resetCurrentColumns];
> +
> +  boardArr = [...newBoard];
> +
>
> Do you need to reset who the current player is?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/prra-codes/peter-arthur-portfolio/pull/8#pullrequestreview-1516259221>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATTKUYBYF6I3SD75QQOCVBDXO2I3VANCNFSM6AAAAAAZ4XXNAE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
